### ISSUE TITLE
visual-fill-column-width and font size changes

### ISSUE THREAD

**Issue by mgrachten** [joostkremers/visual-fill-column#12]
When I increase or decrease the font size (with 'text-scale-increase/'text-scale-decrease) after setting visual-fill-column-width, I would expect the position of the fill column to change according to the font size (since it was specified in character columns), but this is not the case. Is that intentional?

**Comment by joostkremers** [comment]
It's true that the fill column width is based on the character width, but the way text size adjustment is done in Emacs does not change the values that is returned by `window-width` and by `window-margins`, which are used by `visual-line-mode` to calculate how wide the margins should be.

In principle it should be possible to take (relative) text size into account when calculating the margins, but I'm not sure if it's possible to automatically adjust the margins each time the user changes the text size. There is no hook that is run after each text size adjustment, for example.

So I'm not sure if it's worth making such a change. What I think *would* be is to add an option to specify a text size in `writeroom`-enabled buffers. But this would be on the assumption that the user won't change the text size afterwards.

**Comment by joostkremers** [comment]
Oh, sorry, now I'm confusing `writeroom-mode` and `visual-line-mode`...

Of course, the option to change text size would not be part of `visual-line-mode`.

**Comment by cpitclaudel** [comment]
> 

It's true that the fill column width is based on the character width, but the way text size adjustment is done in Emacs does not change the values that is returned by window-width and by window-margins, which are used by visual-line-mode to calculate how wide the margins should be.

I just ran into this issue, and was going to report it after stepping through the code and coming to the same conclusion. Glad to see that you've already diagnosed it :)

> In principle it should be possible to take (relative) text size into account when calculating the margins

I think this would be wonderful

> but I'm not sure if it's possible to automatically adjust the margins each time the user changes the text size. There is no hook that is run after each text size adjustment, for example.

There's no hook, but we can ask the Emacs devs for one :) And in the meantime it should be easy to advise text-scale-adjust, which is very likely to be what people will use for changing the font size. Providing an interactive `writeroom-adjust` (resp. `visual-fill-column-adjust`) function to recompute the margin explicitly would be great too.

In any case, I'm very much +1 on fixing this bug; currently I need to recompute an adjusted number of columns myself in a piece of advice added to `writeroom--enable`.

**Comment by joostkremers** [comment]
> There's no hook, but we can ask the Emacs devs for one :) 

I'll consider that, but I suspect chances are low. I've  kinda worn out my welcome on emacs-devel regarding margin-related questions. :)

> And in the meantime it should be easy to advise `text-scale-adjust`, which is very likely to be what people will use for changing the font size.

That's probably the easier solution. I'm always a bit hesitant about adding advice in a package, though. Advising functions seems more like a user-level configuration. A workable compromise might be to add a remark to the documentation saying how `text-scale-adjust` can be advised.

> Providing an interactive `writeroom-adjust` (resp. `visual-fill-column-adjust`) function to recompute the margin explicitly would be great too.

I guess making `visual-fill-column--adjust-window` interactive would be enough to achieve that.

**Comment by joostkremers** [comment]
I added a note about text size adjustment to the READMEs of `visual-fill-column` and `writeroom-mode` with a code snippet to advice the function `text-scale-adjust` and I provided a function for this purpose, `visual-fill-column-adjust`. Unlike `visual-fill-column--adjust-window`, this function only adjusts the margins if `visual-fill-column` is actually active.

I didn't provide a command to recompute the margins explicitly, because one can use `redraw-display` for that purpose.
